crates/deltalake-core/src/kernel/snapshot/mod.rs

```diff
@@ -0,0 +1,166 @@
+use std::sync::Arc;
+
+use arrow_array::RecordBatch;
+use futures::stream::BoxStream;
+use object_store::path::Path;
+use object_store::ObjectStore;
+
+use self::log_segment::LogSegment;
+use self::replay::ReplayStream;
+use crate::kernel::{actions::ActionType, StructType};
+use crate::table::config::TableConfig;
+use crate::{DeltaResult, DeltaTableConfig};
+
+use super::{Metadata, Protocol};
+
+mod extract;
+mod log_segment;
+mod replay;
+
+/// A snapshot of a Delta table
+pub struct Snapshot {
+    log_segment: LogSegment,
```

Comment:
Will there be any benefit from wrapping this in an Arc? If this will replace the table state struct then it be cloned a bit.

Reply:
We shall see 😆.

IIRC, we tried to avoid cloning the table whenever possible b/c can be quite large, but then eventually caved and made it `Clone`. A goal is to end up in a place where we can do guilt-free clones and (de-)serialization of the `DeltaTable` and / or `Snapshot`. Whether we are already getting there in this PR is another question 😆.